README.md
# MMM-WyzeBridge

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

This module connects to a [Wyze-Bridge](https://github.com/mrlt8/docker-wyze-bridge) server instance and rotate trough registered cameras using [Video.js](https://github.com/videojs/video.js) and [videojs-http-streaming (VHS)](https://github.com/videojs/http-streaming) to be used even within browsers not supporting video tags.

## Example

![](.github/example.png)

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-WyzeBridge',
            config: {
                updateInterval: 30000, // Default time to show next camera (in milliseconds)
                retryDelay: 5000, // Time to ask to wyze-bridge server for cameras updates (in milliseconds)
                controls: false, // If video player should show its controls
                height: 350, // video player height
                width: 700, // video player width
                targetHost: null, // wyze-bridge host to connect (e.g "http://localhost")
                targetPort: null, // wyze-bridge port to connect (e.g 5000)
                animationSpeed: 400, // Animation speed to update DOM
                filter: [], // camera filter (see below)
            }
        }
    ]
}
```

## Configuration options

| Option           | Default | Description                                                                        |
|------------------|---------|------------------------------------------------------------------------------------|
| `updateInterval` | `30000` | *Optional* Default time to show next camera (in milliseconds)                      |
| `retryDelay`     | `5000`  | *Optional* Time to ask to wyze-bridge server for cameras updates (in milliseconds) |
| `controls`       | `false` | *Optional* If video player should show its controls                                |
| `height`         | `350`   | *Optional* video player height                                                     |
| `width`          | `700`   | *Optional* video player width                                                      |
| `targetHost`     | `null`  | *Required* wyze-bridge host to connect (e.g `"http://localhost"`)                  |
| `targetPort`     | `null`  | *Required* wyze-bridge port to connect (e.g 5000)                                  |
| `animationSpeed` | `400`   | *Optional* Animation speed to update DOM                                           |
| `filter`         | `[]`    | *Optional* camera filter (see below)                                               |

### Filtering
Use of *filter* option reduce the amount of cameras to show in player. An empty list means no filtering (use all cameras retrieved). **USE CAREFULLY**.

This option can receive a list of camera names:
```js
    config: {
        ...
        filter: ["cam1", "cam2", ...],
        ...
    }
```
a list of patterns:
```js
    config: {
        ...
        filter: ["/^cam-(1|2).*$/", ...],
        ...
    }
```
or a list of RegExp objects:
```js
    config: {
        ...
        filter: [new RegExp("/^cam-(1|2).*$/"), ...],
        ...
    }
```


CHANGELOG.md
# MMM-WyzeBridge Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Released

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: "MMM-WyzeBridge"
 *
 * By Andrés Vanegas <ukab72106@gmail.com>
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
const Log = require("../../js/logger.js");
const { createProxyMiddleware } = require("http-proxy-middleware");
const axios = require("axios").default;
const nocache = require('nocache');

module.exports = NodeHelper.create({
	name: "MMM-WyzeBridge",
	urlPrefix: null,
	logPrefix: "MMM-WyzeBridge :: ",
	allowedKeys: [
		"connected",
		"enabled",
		"firmware_ver",
		"name_uri",
		"nickname",
		"product_model",
		"status"
	],
	index: 0,
	cameras: [],
	readyState: false,

	start: function () {
		this.config = null;
		this.readyState = false;
		Log.info(this.logPrefix + "Started");
		this.sendNotification("SET_MESSAGE", "LOADING");
	},

	sendNotification(notification, payload) {
		this.sendSocketNotification(this.name + "-" + notification, payload);
	},

	socketNotificationReceived: function (notification, payload) {
		var self = this;
		notification = notification.replace(this.name + "-", "");

		switch (notification) {
			case "SET_CONFIG":
				if (this.config === null) {
					Log.info(this.logPrefix + "Working notification system. Notification:", notification, "payload: ", payload);
					this.config = payload;
					this.urlPrefix = this.config.__protocol + "//localhost:" + this.config.__port + "/" + this.name;
					this.setProxy();
					this.swapCamera();
				} else {
					this.showCamera();
				}

				break;
		}
	},

	// this you can create extra routes for your module
	setProxy: function () {
		var self = this;
		this.expressApp.set("etag", false);
		this.expressApp.use("/" + this.name + "/proxy/*",
			nocache(),
			createProxyMiddleware({
				target: this.config.targetHost + ":" + this.config.targetPort, // target host with the same base path
				changeOrigin: true, // needed for virtual hosted sites
				pathRewrite: function (path, _) {
					return path.replace(new RegExp("^/" + self.name + "/proxy/"), "/");
				},
			})
		);

		this.expressApp.use("/" + this.name + "/stream/*",
			nocache(),
			createProxyMiddleware({
				target: this.config.targetHost + ":8888", // target host with the same base path
				changeOrigin: true, // needed for virtual hosted sites
				pathRewrite: function (path, _) {
					return path.replace(new RegExp("^/" + self.name + "/stream/"), "/");
				},
			})
		);

		this.getCameras();
	},

	processCameras: function (data) {
		var self = this;
		if (typeof data !== "object" || !data.hasOwnProperty("cameras")) {
			data = { cameras: {} };
		}

		var cameras = [];
		for (var cameraName in data.cameras) {
			const __canShowCamera = (
				// All cameras should be shown or
				this.config.filter == 0
				// Camera match exactly a filter as string
				|| this.config.filter.includes(cameraName)
				// Camera match a valid pattern
				|| this.config.filter.map(p => new RegExp(p, "gi")).some(p => cameraName.match(p))
			);
			if (__canShowCamera) {
				var cameraData = {
					image_url: "/proxy/" + data.cameras[cameraName].snapshot_url,
					video_url: "/stream/" + data.cameras[cameraName].name_uri + "/stream.m3u8",
				};
				for (camAttribute of this.allowedKeys) {
					cameraData[camAttribute] = data.cameras[cameraName][camAttribute];
				}
				cameras.push(cameraData);
			}
		}
		cameras.sort(function (a, b) {
			if (a.nickname < b.nickname) { return -1; }
			if (a.nickname > b.nickname) { return 1; }
			return 0;
		});

		camerasReceived = cameras.map(x => x.nickname);
		camerasAlreadyDetected = this.cameras.map(x => x.nickname);

		var removedCameras = camerasAlreadyDetected.filter(x => !camerasReceived.includes(x));
		var newCameras = camerasReceived.filter(x => !camerasAlreadyDetected.includes(x));
		if (newCameras.length + removedCameras.length > 0) {
			Log.info(self.logPrefix + "Changes received in cameras");
			this.cameras = cameras;
			this.index = this.index > (this.cameras.length - 1) ? 0 : this.index;
			newCameras.forEach(x => Log.log(self.logPrefix + x + " camera detected"));
			removedCameras.forEach(x => Log.log(self.logPrefix + x + " camera removed"));
		}
	},

	changeIndex: function () {
		if (this.readyState && this.cameras.length > 0) {
			if (this.index < (this.cameras.length - 1)) {
				this.index++;
			} else {
				this.index = 0;
			}
		}
	},

	showCamera: function () {
		var self = this;
		if (!this.readyState) {
			Log.info(this.logPrefix + "Still loading");
			this.sendNotification("SET_MESSAGE", "LOADING");
		}
		else if (this.cameras.length == 0) {
			Log.info(this.logPrefix + "No cameras found. skipping");
			this.sendNotification("SET_MESSAGE", "NO_CAMS");
		}
		else {
			Log.info(this.logPrefix + "Showing camera " + this.cameras[this.index].nickname);
			this.sendNotification("SET_CAMERA", this.cameras[this.index]);
		}
	},

	swapCamera: function () {
		var self = this;
		this.showCamera();
		this.changeIndex();

		setTimeout(function () {
			self.swapCamera();
		}, this.readyState ? this.config.updateInterval : this.config.retryDelay);
	},

	// Test another function
	getCameras: function () {
		var self = this;

		Log.log(self.logPrefix + "Requesting cameras...");
		axios.get(this.urlPrefix + "/proxy/cameras")
			.then(function (response) {
				self.processCameras(response.data);
			})
			.catch(function (error) {
				self.cameras = [];
				Log.error(self.logPrefix + error.message);
			})
			.then(function () {
				self.readyState = true;
				self.sendNotification("READY_STATE", self.readyState);
				self.sendNotification("CAMERAS_UPDATED", self.cameras.length);

				Log.log(self.logPrefix + "Request cameras finished");
				setTimeout(function () { self.getCameras(); }, self.config.retryDelay);
			});
	}
});


MMM-WyzeBridge.js
/* global Module */

/* Magic Mirror
 * Module: MMM-WyzeBridge
 *
 * By Andrés Vanegas <ukab72106@gmail.com>
 * MIT Licensed.
 */
Module.register("MMM-WyzeBridge", {
	/**
	 * @member {Object} defaults - Defines the default config values.
	 * @property {int} updateInterval Default time to show next camera (in milliseconds). Defaults to 30000.
	 * @property {int} retryDelay Time to ask to wyze-bridge server for cameras updates (in milliseconds). Defaults to 5000.
	 * @property {boolean} controls If video player should show its controls. Defaults to false.
	 * @property {int} height video player height. Defaults to 350.
	 * @property {int} width video player width. Defaults to 700.
	 * @property {string} targetHost wyze-bridge host to connect (e.g http://localhost). Defaults to null.
	 * @property {int} targetPort wyze-bridge port to connect (e.g 5000). Defaults to null.
	 * @property {int} animationSpeed Animation speed to update DOM. Defaults to 400.
	 * @property {str[]|RegExp[]} filter camera filters (can be a string representing name, pattern or a regex). Defaults to [] (empty array).
	*/
	defaults: {
		updateInterval: 30000,
		retryDelay: 5000,
		controls: false,
		height: 350,
		width: 700,
		targetHost: null,
		targetPort: null,
		animationSpeed: 400,
		filter: [],
	},

	// Required version of MagicMirror
	requiresVersion: "2.1.0",

	// Placeholders
	wrapper: null,
	message: null,
	messageWrapper: null,
	playerWrapper: null,
	player: null,
	cameras: 0,

	// Overrides start method
	start: function () {
		this.config = {
			...this.defaults,
			...this.config,
		};
		this.message = this.translate("LOADING");

		this.updateDom(this.config.animationSpeed);
		this.sendNotification("SET_CONFIG", {
			...this.config,
			__protocol: window.location.protocol,
			__port: window.location.port,
		});
	},

	/**
	 * Notification send helper method
	 * @param {string} notification notification type
	 * @param {any} payload notification payload
	 */
	sendNotification(notification, payload) {
		this.sendSocketNotification(this.name + "-" + notification, payload);
	},

	// Override socket notification received method
	socketNotificationReceived: function (notification, payload) {
		switch (notification.replace(this.name + "-", "")) {
			case "CAMERAS_UPDATED":
				this.cameras = payload;
				break;
			case "SET_MESSAGE":
				this.message = this.translate(payload)
				this.updateDom(this.config.animationSpeed);
				break;
			case "SET_CAMERA":
				this.message = null;
				this.updateDom(this.config.animationSpeed);
				this.swapSource(payload);
				break;
		}
	},

	/**
	 * Change current camera and activate player
	 * @param {object} camera camera data to be show
	 */
	swapSource: function (camera) {
		Log.log("Showing camera " + camera.nickname);
		if (this.player === null) {
			this.showPlayer();
		}
		// this.player.poster("/" + this.name + camera.image_url);
		this.player.src({
			src: "/" + this.name + camera.video_url,
			type: 'application/x-mpegURL',
		});
		this.player.play();
	},

	/**
	 * Clears an attribute from this instance
	 * @param {string} nodename attribute to be clear
	 */
	clearNode(nodename) {
		try {
			switch (nodename) {
				case 'player':
					if (this.player !== null) {
						this.player.dispose();
					}
					break;
				default:
					if (this[nodename] !== null) {
						this[nodename].parentNode.removeChild(this[nodename])
					}
			}
		} catch (e) { }

		if (this.hasOwnProperty(nodename)) {
			this[nodename] = null;
		}
	},

	/**
	 * Show message in wrapper and hide the player
	 */
	showMessage() {
		this.clearNode('player');
		this.clearNode('playerWrapper');

		if (this.messageWrapper === null) {
			this.messageWrapper = document.createElement("div");
			this.messageWrapper.classList.add("message-container");
			this.messageWrapper.style.width = this.config.width + "px";
			this.messageWrapper.style.height = this.config.height + "px";
			this.wrapper.appendChild(this.messageWrapper);
		}
		this.messageWrapper.innerHTML = this.message;
	},

	/**
	 * Show player and hide the message
	 */
	showPlayer() {
		this.clearNode('messageWrapper');

		if (this.playerWrapper === null) {
			this.playerWrapper = document.createElement("video-js");
			this.playerWrapper.classList.add("player_" + this.name);
			this.playerWrapper.setAttribute("id", "player_" + this.identifier);
			this.wrapper.appendChild(this.playerWrapper);
		}

		if (this.player === null && this.playerWrapper.offsetParent !== null) {
			try {
				var player = videojs(this.playerWrapper, {
					autoplay: false,
					controls: this.config.controls,
					muted: "muted",
					preload: "none",
					width: this.config.width,
					height: this.config.height,
					fluid: true,
					liveui: true,
					loadingSpinner: false,
				});
				this.player = player;
			} catch (e) {
				this.clearNode('player');
				Log.error(e);
			}
		}
	},

	// Override function to retrieve DOM elements
	getDom: function () {
		if (this.wrapper === null) {
			this.wrapper = document.createElement("div");
			this.wrapper.classList.add("wrapper_" + this.name);
			this.wrapper.style.width = this.config.width + "px";
			this.wrapper.style.height = this.config.height + "px";
		}

		if (this.message !== null) {
			this.showMessage();
		} else {
			this.showPlayer();
		}

		return this.wrapper;
	},

	// Load scripts
	getScripts: function () {
		const __lang = this.config.lang || this.language || "en";
		return [
			this.file("js/video.min.js"),
			this.file("js/lang/" + (__lang) + ".js"),
			this.file("js/videojs-http-streaming.min.js"),
		];
	},

	// Load stylesheets
	getStyles: function () {
		return [
			this.file("css/video-js.css"),
			this.name + ".css",
		];
	},

	// Load translations files
	getTranslations: function () {
		//FIXME: This can be load a one file javascript definition
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},
});
